--- openssl-0.9.8k.orig/crypto/err/err_all.c	2008-11-24 08:14:14.000000000 -0800
+++ openssl-0.9.8k/crypto/err/err_all.c	2009-09-29 14:23:15.000000000 -0700
@@ -146,7 +146,9 @@ void ERR_load_crypto_strings(void)
 #ifndef OPENSSL_NO_ENGINE
 	ERR_load_ENGINE_strings();
 #endif
+#ifndef OPENSSL_NO_OCSP
 	ERR_load_OCSP_strings();
+#endif
 	ERR_load_UI_strings();
 #ifdef OPENSSL_FIPS
 	ERR_load_FIPS_strings();
