--- openssl-1.0.0.orig/apps/openssl.c	2009-10-04 09:43:21.000000000 -0700
+++ openssl-1.0.0/apps/openssl.c	2010-05-18 14:05:14.000000000 -0700
@@ -275,8 +275,10 @@ int main(int Argc, char *Argv[])
 		if (ERR_GET_REASON(ERR_peek_last_error())
 		    == CONF_R_NO_SUCH_FILE)
 			{
+#if 0 /* ANDROID */
 			BIO_printf(bio_err,
 				   "WARNING: can't open config file: %s\n",p);
+#endif
 			ERR_clear_error();
 			NCONF_free(config);
 			config = NULL;
--- openssl-1.0.0.orig/apps/progs.h	2009-06-30 08:08:38.000000000 -0700
+++ openssl-1.0.0/apps/progs.h	2010-05-18 14:05:38.000000000 -0700
@@ -43,7 +43,9 @@ extern int spkac_main(int argc,char *arg
 extern int smime_main(int argc,char *argv[]);
 extern int rand_main(int argc,char *argv[]);
 extern int engine_main(int argc,char *argv[]);
+#ifndef OPENSSL_NO_OCSP
 extern int ocsp_main(int argc,char *argv[]);
+#endif
 extern int prime_main(int argc,char *argv[]);
 extern int ts_main(int argc,char *argv[]);
 
@@ -107,7 +109,9 @@ FUNCTION functions[] = {
 #endif
 	{FUNC_TYPE_GENERAL,"genpkey",genpkey_main},
 #if !defined(OPENSSL_NO_SOCK) && !(defined(OPENSSL_NO_SSL2) && defined(OPENSSL_NO_SSL3))
-	{FUNC_TYPE_GENERAL,"s_server",s_server_main},
+#if 0 /* ANDROID */
+        {FUNC_TYPE_GENERAL,"s_server",s_server_main},
+#endif
 #endif
 #if !defined(OPENSSL_NO_SOCK) && !(defined(OPENSSL_NO_SSL2) && defined(OPENSSL_NO_SSL3))
 	{FUNC_TYPE_GENERAL,"s_client",s_client_main},
@@ -133,9 +137,11 @@ FUNCTION functions[] = {
 	{FUNC_TYPE_GENERAL,"pkcs12",pkcs12_main},
 #endif
 	{FUNC_TYPE_GENERAL,"pkcs8",pkcs8_main},
+#if 0 /* ANDROID */
 	{FUNC_TYPE_GENERAL,"pkey",pkey_main},
 	{FUNC_TYPE_GENERAL,"pkeyparam",pkeyparam_main},
 	{FUNC_TYPE_GENERAL,"pkeyutl",pkeyutl_main},
+#endif
 	{FUNC_TYPE_GENERAL,"spkac",spkac_main},
 	{FUNC_TYPE_GENERAL,"smime",smime_main},
 	{FUNC_TYPE_GENERAL,"rand",rand_main},
@@ -146,7 +152,9 @@ FUNCTION functions[] = {
 	{FUNC_TYPE_GENERAL,"ocsp",ocsp_main},
 #endif
 	{FUNC_TYPE_GENERAL,"prime",prime_main},
+#if 0 /* ANDROID */
 	{FUNC_TYPE_GENERAL,"ts",ts_main},
+#endif
 #ifndef OPENSSL_NO_MD2
 	{FUNC_TYPE_MD,"md2",dgst_main},
 #endif
--- openssl-1.0.0.orig/apps/speed.c	2010-03-03 11:56:17.000000000 -0800
+++ openssl-1.0.0/apps/speed.c	2010-05-18 14:05:57.000000000 -0700
@@ -1718,6 +1718,7 @@ int MAIN(int argc, char **argv)
 			}
 		}
 
+#if 0 /* ANDROID */
 	if (doit[D_IGE_128_AES])
 		{
 		for (j=0; j<SIZE_NUM; j++)
@@ -1763,6 +1764,7 @@ int MAIN(int argc, char **argv)
 
 
 #endif
+#endif
 #ifndef OPENSSL_NO_CAMELLIA
 	if (doit[D_CBC_128_CML])
 		{
--- openssl-1.0.0.orig/crypto/ui/ui_openssl.c	2009-10-04 09:43:21.000000000 -0700
+++ openssl-1.0.0/crypto/ui/ui_openssl.c	2010-05-18 13:36:26.000000000 -0700
@@ -184,7 +184,7 @@
 # undef  SGTTY
 #endif
 
-#if defined(linux) && !defined(TERMIO)
+#if defined(linux) && !defined(TERMIO) && !defined(__ANDROID__)
 # undef  TERMIOS
 # define TERMIO
 # undef  SGTTY
--- openssl-1.0.0.orig/apps/s_client.c	2009-12-16 12:28:28.000000000 -0800
+++ openssl-1.0.0/apps/s_client.c	2010-04-14 14:25:09.000000000 -0700
@@ -216,6 +216,9 @@ static int c_ign_eof=0;
 /* Default PSK identity and key */
 static char *psk_identity="Client_identity";
 /*char *psk_key=NULL;  by default PSK is not used */
+#if 1 /* ANDROID */
+char *psk_key=NULL;
+#endif
 
 static unsigned int psk_client_cb(SSL *ssl, const char *hint, char *identity,
 	unsigned int max_identity_len, unsigned char *psk,
